File: script/compress_image.py
#!/usr/bin/env python3

"""compress_image.py

Small utility to compress PNG and JPEG images in a directory tree using
external tools (`pngquant` for PNG, `jpegoptim` for JPEG). Designed to be
run after building a static site (default target directory is `site`).

Behavior notes:
- PNG files are compressed with `pngquant` producing a temporary file which
  then replaces the original.
- JPEG files are processed with `jpegoptim` and overwritten in-place.
- The script runs compressions in parallel using a thread pool controlled by
  `THREADS_NUM`.
- This script modifies files in place; create backups or run on a copy if you
  want to preserve originals.
"""

import subprocess
from pathlib import Path
import os
import argparse

# Compression quality settings
JPG_QUALITY = 80       # JPEG quality (0-100), lower => smaller files
PNG_QUALITY = "60-80"  # pngquant quality range (min-max as percentages)
THREADS_NUM = 4        # Number of worker threads for parallel processing


def compress_image_once(file_path: Path) -> None:
    """Compress a single image file in-place.

    The function chooses the compression tool based on file extension:
    - .png -> pngquant (writes a temp file then replaces original)
    - .jpg/.jpeg -> jpegoptim (overwrites original)

    Errors are caught and printed; they do not stop processing other files.

    Args:
        file_path: Path object pointing to the image file to compress.
    """
    try:
        suffix = file_path.suffix.lower()

        if suffix == ".png":
            temp_file = file_path.with_suffix(".tmp.png")
            # Run pngquant to compress PNG into a temporary file
            subprocess.run([
                "pngquant", "--force",
                "--quality", PNG_QUALITY,
                "--output", str(temp_file),
                str(file_path)
            ], check=True)
            # Replace original with compressed file
            os.replace(temp_file, file_path)

        elif suffix in (".jpg", ".jpeg"):
            # Run jpegoptim to compress and strip metadata
            subprocess.run([
                "jpegoptim",
                "--max=" + str(JPG_QUALITY),
                "--strip-all",
                "--overwrite",
                str(file_path)
            ], check=True)

        # Log resulting file size in KB
        print(f"[compress] [info] {file_path} ({file_path.stat().st_size/1024:.1f} KB)")

    except Exception as e:
        # Print the error and continue
        print(f"[compress] [fail] {file_path}: {e}")


def compress_image(root_dir: str) -> None:
    """Recursively find PNG/JPEG images under `root_dir` and compress them.

    Files are discovered using pathlib's `rglob`. Compression tasks are
    dispatched to a thread pool for parallel execution.

    Args:
        root_dir: Root directory to search for image files.
    """
    image_exts = (".png", ".jpg", ".jpeg")
    image_files = []

    # Collect all supported image files recursively
    for ext in image_exts:
        image_files.extend(Path(root_dir).rglob(f"*{ext}"))

    # Compress images in parallel using a thread pool
    from concurrent.futures import ThreadPoolExecutor
    with ThreadPoolExecutor(max_workers=THREADS_NUM) as executor:
        executor.map(compress_image_once, image_files)


if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        description="Compress PNG and JPEG images in a directory tree (in-place)."
    )
    parser.add_argument(
        "dir",
        type=str,
        nargs="?",
        default="site",
        help="Directory to compress images in (default: 'site')",
    )
    args = parser.parse_args()

    # Ensure target directory exists
    if not Path(args.dir).exists():
        print(f"[compress] [fail] directory '{args.dir}' does not exist!")
        raise SystemExit(1)

    compress_image(args.dir)

File: script/compress_image.js
// compress_image.js
// Utility script to find raster images under a directory, convert them to
// WebP using `sharp`, and remove the original files. Intended for use after
// building a static site (`site` by default).
//
// Notes:
// - This script is destructive: original PNG/JPG files are deleted after
//   successful conversion to WebP. Run on a copy or commit changes beforehand
//   if you want to keep originals.
// - Concurrency is controlled by THREADS_NUM to avoid excessive parallelism.
// - Output files are written alongside originals with `.webp` extension.

import { promises as fs } from "fs";
import path from "path";
import sharp from "sharp";

// Compression quality settings
// IMG_QUALITY: 0-100 (higher => better quality and larger file size)
const IMG_QUALITY = 80;
// THREADS_NUM: number of concurrent workers compressing images
const THREADS_NUM = 4;

// Recursively find image files under `dir`.
// Returns an array of absolute file paths for .png, .jpg, .jpeg files.
async function findImages(dir) {
    let results = [];
    const entries = await fs.readdir(dir, { withFileTypes: true });

    for (const entry of entries) {
        const fullPath = path.join(dir, entry.name);

        if (entry.isDirectory()) {
            // Recurse into subdirectories
            results = results.concat(await findImages(fullPath));
        }
        else {
            const ext = path.extname(entry.name).toLowerCase();
            if ([".png", ".jpg", ".jpeg"].includes(ext)) {
                results.push(fullPath);
            }
        }
    }

    return results;
}

// Run an array of task functions with a maximum concurrency of `limit`.
// Each task is expected to be an async function that performs one unit of work.
async function runWithConcurrency(tasks, limit) {
    const queue = [];
    let i = 0;

    async function worker() {
        while (i < tasks.length) {
            const currentIndex = i++;
            // Execute the task and wait for it to finish before taking next
            await tasks[currentIndex]();
        }
    }

    // Start `limit` workers in parallel
    for (let j = 0; j < limit; j++) {
        queue.push(worker());
    }

    await Promise.all(queue);
}

// Convert a single image file to WebP and remove the original file on success.
// This function is robust to unsupported file extensions and logs failures
// without throwing, so the concurrency loop can continue processing others.
async function compressImageOnce(filePath) {
    try {
        const ext = path.extname(filePath).toLowerCase();

        if (![".png", ".jpg", ".jpeg"].includes(ext)) return;

        // Replace original extension with .webp
        const outputPath = filePath.replace(/\.(png|jpg|jpeg)$/i, ".webp");

        // Use sharp to write WebP with configured quality
        await sharp(filePath)
            .webp({ quality: IMG_QUALITY })
            .toFile(outputPath);

        // Remove the original file after successful conversion
        await fs.unlink(filePath);

        console.log(`[compress] [info] ${filePath} -> ${outputPath}`);
    }
    catch (err) {
        // Log error and continue processing other images
        console.error(`[compress] [fail] ${filePath}: ${err.message}`);
    }
}

// Find images under `rootDir` and compress them with limited concurrency.
async function compressImage(rootDir) {
    const images = await findImages(rootDir);
    console.log(`[compress] [info] found ${images.length} images`);

    // Wrap each file path into a task function so runWithConcurrency can call them
    const tasks = images.map((file) => () => compressImageOnce(file));
    await runWithConcurrency(tasks, THREADS_NUM);
    console.log("[compress] [info] done!");
}

// Entry point: accept an optional directory argument (defaults to `site`).
// Validates the directory exists and then starts compression.
async function main() {
    const args_dir = process.argv[2] || "site";

    try {
        await fs.access(args_dir);
    }
    catch {
        console.error(`[compress] [fail] directory "${args_dir}" does not exist!`);
        process.exit(1);
    }

    // Note: do not await here so the top-level promise is allowed to settle
    // if you prefer to catch all errors, you can `await` compressImage(args_dir)
    compressImage(args_dir);
}

main();
